sgkit/stats/genedrop.py

```diff
@@ -0,0 +1,344 @@
+from typing import Hashable, Optional
+
+import dask.array as da
+import numpy as np
+import xarray as xr
+from typing_extensions import Literal
+from xarray import Dataset
+
+from sgkit import variables
+from sgkit.accelerate import numba_guvectorize, numba_jit
+from sgkit.typing import ArrayLike
+from sgkit.utils import (
+    conditional_merge_datasets,
+    create_dataset,
+    define_variable_if_absent,
+)
+
+from .pedigree import (
+    _compress_hamilton_kerr_parameters,
+    parent_indices,
+    topological_argsort,
+)
+
+EST_DIPLOID = "diploid"
+EST_HAMILTON_KERR = "Hamilton-Kerr"
+
+
+@numba_guvectorize(  # type: ignore
+    [
+        "void(int8[:,:], int64[:,:], uint32, int8[:,:])",
+        "void(int16[:,:], int64[:,:], uint32, int16[:,:])",
+        "void(int32[:,:], int64[:,:], uint32, int32[:,:])",
+        "void(int64[:,:], int64[:,:], uint32, int64[:,:])",
+    ],
+    "(n,k),(n,p),()->(n,k)",
+)
+def genedrop_diploid(
+    genotypes: ArrayLike,
+    parent: ArrayLike,
+    seed: ArrayLike,
+    out: ArrayLike,
+) -> None:  # pragma: no cover
+    n_sample, n_parent = parent.shape
+    _, ploidy = genotypes.shape
+    if n_parent != 2:
+        raise ValueError("The parents dimension must be length 2")
+    if ploidy != 2:
+        raise ValueError("Genotypes are not diploid")
+    order = topological_argsort(parent)
+    np.random.seed(seed)
+    for i in range(n_sample):
+        t = order[i]
+        unknown_parent = 0
+        for j in range(n_parent):
+            p = parent[t, j]
+            if p < 0:
+                # founder
+                unknown_parent += 1
+            else:
+                idx = np.random.randint(2)
+                out[t, j] = out[p, idx]
+        if unknown_parent == 1:
+            raise ValueError("Pedigree contains half-founders")
+        elif unknown_parent == 2:
+            # copy founder
+            out[t, 0] = genotypes[t, 0]
+            out[t, 1] = genotypes[t, 1]
+
+
+@numba_jit(nogil=True)
+def _random_gamete_Hamilton_Kerr(
+    genotype: ArrayLike, ploidy: int, tau: int, lambda_: float
+) -> ArrayLike:
+    if ploidy < len(genotype):
```

Comment:
Probably but it gets a little complex. These checks involve evaluation of the relevant arrays. So, an independent validation function could cause an array to be evaluated twice. This may be avoided if the separate validation function is somehow added to the task graph as a dependency of the compute function.